spec/src/main/resources/META-INF/smithy/bsp/extensions/rust.smithy

```diff
@@ -0,0 +1,197 @@
+$version: "2"
+
+namespace bsp.rust
+
+use bsp#BuildTargetData
+use bsp#BuildTargetIdentifiers
+use jsonrpc#data
+use jsonrpc#dataKind
+use jsonrpc#jsonRPC
+use jsonrpc#jsonRequest
+
+// TODO: check correctness. Refer to original Rust extension implementation in .xtend/java/scala.
+
+@jsonRPC
+service RustBuildServer {
+    operations: [
+        RustWorkspace
+        RustToolchain
+    ]
+}
+
+@jsonRequest("buildTarget/rustWorkspace")
+operation RustWorkspace {
+    input: RustWorkspaceParams
+    output: RustWorkspaceResult
+}
+
+structure RustWorkspaceParams {
+    @required
+    targets: BuildTargetIdentifiers
+}
+
+structure RustWorkspaceResult {
+    @required
+    packages: RustPackages
+    @required
+    rawDependencies: RustRawDependencies
+    @required
+    dependencies: RustDependencies
+}
+
+list RustPackages {
+    member: RustPackage
+}
+
+structure RustPackage {
+    @required
+    id: String
+    version: String
+    origin: String
+    edition: String
+    source: String
+    @required
+    targets: RustTargets
+    @required
+    allTargets: RustTargets
+    @required
+    features: RustFeatures
+    @required
+    enabledFeatures: Strings
+    cfgOptions: RustCfgOptions
+    @required
+    env: RustEnvDatas
+    outDirUrl: String
+    procMacroArtifact: RustProcMacroArtifact
+}
+
+list RustTargets {
+    member: RustTarget
+}
+
+@dataKind(kind: "rust", extends: BuildTargetData)
+structure RustTarget {
+    @required
+    name: String
+    @required
+    crateRootUrl: String
+    @required
+    packageRootUrl: String
+    @required
+    kind: String
+    edition: String
+    doctest: Boolean
+    requiredFeatures: Strings
+}
+
+list RustFeatures {
+    member: RustFeature
+}
+
+structure RustFeature {
+    @required
+    name: String
+    deps: Strings
+}
+
+list Strings {
+    member: String
+}
+
+structure RustCfgOptions {
+    keyValueOptions: KeyValueOptions
+    nameOptions: Strings
+}
+
+map KeyValueOptions {
+    key: String
+    value: Strings
+}
+
+map RustEnvDatas {
+    key: String
+    value: String
+}
+
+structure RustProcMacroArtifact {
+    path: String
+    // TODO: we don't need hash. It is calculated by IntelliJ-Rust
+}
+
+map RustRawDependencies {
+    key: String // PackageId
+    value: RustRawDependency
+}
+
+structure RustRawDependency {
+    @required
+    name: String
+    rename: String
```

Comment:
what is it?

Reply:
Name to which this dependency is renamed when declared in Cargo.toml

Reply:
After obtaining information about the project (either from BSP or Cargo itself), `intellij-rust` creates their own structures with packages and dependencies. Then, the rename is used as the `cargo feature dependency package name` (and if it is null, the package name is used). This field is later used in the plugin for example for making feature graph and finding needed dependencies. In general, the `rename` field in Cargo allows to specify an alternative name for a dependency to use in a code, regardless of how it’s published (helpful for example if multiple dependencies have conflicting names)

Reply:
@julia-podrazka could you extract the intellij-agnostic part of this comment to the docs?

Reply:
Done